### PR TITLE
Add type hints for mmcv/ops

### PR DESCRIPTION
**Pull request by triple-Mu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add type hints for mmcv/ops

## Modification

mmcv/ops

## BC-breaking (Optional)

No

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by zhouzaida** [comment]
Hi @triple-Mu , lint failed.

**Comment by zhouzaida** [comment]
Need to add type hints for return values.

File: mmcv/ops/roi_align.py [hunk 1]
```diff
@@ -14,8 +16,9 @@
 class RoIAlignFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
-                 pool_mode, aligned):
+    def symbolic(g, input: torch.Tensor, rois: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/roi_align.py:19]:
```suggestion
    def symbolic(g: torch._C.graph, input: torch.Tensor, rois: torch.Tensor,
```

File: mmcv/ops/roi_align_rotated.py [hunk 2]
```diff
@@ -12,8 +15,9 @@
 class RoIAlignRotatedFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
-                 aligned, clockwise):
+    def symbolic(g, input: torch.Tensor, rois: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/roi_align_rotated.py:18]: same as above

File: mmcv/ops/roi_pool.py [hunk 3]
```diff
@@ -49,7 +56,7 @@ def forward(ctx, input, rois, output_size, spatial_scale=1.0):
 
     @staticmethod
     @once_differentiable
-    def backward(ctx, grad_output):
+    def backward(ctx: Any, grad_output: torch.Tensor):
```
Comment (HAOCHENYE) [inline mmcv/ops/roi_pool.py:59]: Type hints for output

File: mmcv/ops/roi_align_rotated.py [hunk 4]
```diff
@@ -71,7 +75,7 @@ def forward(ctx,
         return output
 
     @staticmethod
-    def backward(ctx, grad_output):
+    def backward(ctx: Any, grad_output: torch.Tensor):
```
Comment (HAOCHENYE) [inline mmcv/ops/roi_align_rotated.py:78]: Type hints for output

File: mmcv/ops/roi_pool.py [hunk 5]
```diff
@@ -14,7 +16,8 @@
 class RoIPoolFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, rois, output_size, spatial_scale):
+    def symbolic(g, input: torch.Tensor, rois: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/roi_pool.py:19]: same as above

File: mmcv/ops/roiaware_pool3d.py [hunk 6]
```diff
@@ -108,7 +115,7 @@ def forward(ctx, rois, pts, pts_feature, out_size, max_pts_per_voxel,
         return pooled_features
 
     @staticmethod
-    def backward(ctx, grad_out):
+    def backward(ctx: Any, grad_out: torch.Tensor):
```
Comment (HAOCHENYE) [inline mmcv/ops/roiaware_pool3d.py:118]: Type hints for output

File: mmcv/ops/roipoint_pool3d.py [hunk 7]
```diff
@@ -73,5 +81,5 @@ def forward(ctx, points, point_features, boxes3d, num_sampled_points=512):
         return pooled_features, pooled_empty_flag
 
     @staticmethod
-    def backward(ctx, grad_out):
+    def backward(ctx: Any, grad_out: torch.Tensor) -> None:
```
Comment (HAOCHENYE) [inline mmcv/ops/roipoint_pool3d.py:84]: The signature for backward should be consistent with the previous.

File: mmcv/ops/rotated_feature_align.py [hunk 8]
```diff
@@ -21,7 +23,8 @@ class RotatedFeatureAlignFunction(Function):
     """
 
     @staticmethod
-    def symbolic(g, features, best_rbboxes, spatial_scale, points):
+    def symbolic(g, features: torch.Tensor, best_rbboxes: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/rotated_feature_align.py:26]: same as above

File: mmcv/ops/rotated_feature_align.py [hunk 9]
```diff
@@ -84,9 +88,9 @@ def backward(ctx, grad_output):
         return grad_input, None, None, None
 
 
-def rotated_feature_align(features,
-                          best_rbboxes,
-                          spatial_scale=1 / 8,
-                          points=1):
+def rotated_feature_align(features: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/rotated_feature_align.py:91]: Type hints for output

File: mmcv/ops/sparse_modules.py [hunk 10]
```diff
@@ -13,24 +13,26 @@
...
+from typing import Any, List, Optional, Union
 
 import torch
 from torch import nn
 
 from .sparse_structure import SparseConvTensor
 
 
-def is_spconv_module(module):
+def is_spconv_module(module: tuple) -> bool:
```
Comment (HAOCHENYE) [inline mmcv/ops/sparse_modules.py:24]: Why tuple

File: mmcv/ops/sparse_modules.py [hunk 11]
```diff
@@ -13,24 +13,26 @@
...
     spconv_modules = (SparseModule, )
     return isinstance(module, spconv_modules)
 
 
-def is_sparse_conv(module):
+def is_sparse_conv(module: Any) -> bool:
```
Comment (HAOCHENYE) [inline mmcv/ops/sparse_modules.py:29]: Maybe module is a `nn.Module` instance?

File: mmcv/ops/sparse_modules.py [hunk 12]
```diff
@@ -101,7 +103,7 @@ def __init__(self, *args, **kwargs):
             self.add_module(name, module)
         self._sparity_dict = {}
 
-    def __getitem__(self, idx):
+    def __getitem__(self, idx: int):
```
Comment (HAOCHENYE) [inline mmcv/ops/sparse_modules.py:106]: type hints for return value.

File: mmcv/ops/sync_bn.py [hunk 13]
```diff
@@ -19,8 +21,10 @@
 class SyncBatchNormFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, running_mean, running_var, weight, bias, momentum,
-                 eps, group, group_size, stats_mode):
+    def symbolic(g, input: torch.Tensor, running_mean: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/sync_bn.py:24]: same as above

File: mmcv/ops/upfirdn2d.py [hunk 14]
```diff
@@ -108,8 +110,11 @@
 class UpFirDn2dBackward(Function):
 
     @staticmethod
-    def forward(ctx, grad_output, kernel, grad_kernel, up, down, pad, g_pad,
-                in_size, out_size):
+    def forward(ctx: Any, grad_output: torch.Tensor, kernel: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/upfirdn2d.py:113]: type hints for return value.

File: mmcv/ops/psa_mask.py [hunk 15]
```diff
@@ -12,15 +15,15 @@
...
+    def symbolic(g, input: torch.Tensor, psa_type: str, mask_size: int):
         return g.op(
             'mmcv::MMCVPSAMask',
             input,
             psa_type_i=psa_type,
             mask_size_i=mask_size)
 
     @staticmethod
-    def forward(ctx, input, psa_type, mask_size):
+    def forward(ctx, input: torch.Tensor, psa_type: str, mask_size: int):
```
Comment (HAOCHENYE) [inline mmcv/ops/psa_mask.py:26]: The type hints of return value.

File: mmcv/ops/psa_mask.py [hunk 16]
```diff
@@ -82,7 +85,7 @@ def __init__(self, psa_type, mask_size=None):
         self.mask_size = mask_size
         self.psa_type = psa_type
 
-    def forward(self, input):
+    def forward(self, input: torch.Tensor):
```
Comment (HAOCHENYE) [inline mmcv/ops/psa_mask.py:88]: The type hints of return value.

File: mmcv/ops/riroi_align_rotated.py [hunk 17]
```diff
@@ -124,7 +127,7 @@ def __init__(self,
         self.num_orientations = int(num_orientations)
         self.clockwise = clockwise
 
-    def forward(self, features, rois):
+    def forward(self, features: torch.Tensor, rois: torch.Tensor):
```
Comment (HAOCHENYE) [inline mmcv/ops/riroi_align_rotated.py:130]: The type hints of return value.

File: mmcv/ops/sync_bn.py [hunk 18]
```diff
@@ -35,8 +39,10 @@ def symbolic(g, input, running_mean, running_var, weight, bias, momentum,
             stats_mode=stats_mode)
 
     @staticmethod
-    def forward(self, input, running_mean, running_var, weight, bias, momentum,
-                eps, group, group_size, stats_mode):
+    def forward(self, input: torch.Tensor, running_mean: torch.Tensor,
```
Comment (HAOCHENYE) [inline mmcv/ops/sync_bn.py:42]: The type hints of return value.

**Review by HAOCHENYE — APPROVED** [review]

File: mmcv/ops/rotated_feature_align.py (hunk 9 above)
Comment (teamwong111) [reply mmcv/ops/rotated_feature_align.py:92]: should be torch.Tensor.

**Review by teamwong111 — APPROVED** [review]

**Review by imabackstabber — APPROVED** [review]

File: mmcv/ops/riroi_align_rotated.py [hunk 19]
```diff
@@ -11,14 +14,14 @@
...
-                out_size,
-                spatial_scale,
-                num_samples=0,
-                num_orientations=8,
-                clockwise=False):
+    def forward(ctx: Any,
+                features: torch.Tensor,
+                rois: torch.Tensor,
+                out_size: Union[int, tuple],
+                spatial_scale: int,
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:21]:
```suggestion
                spatial_scale: float,
```

File: mmcv/ops/riroi_align_rotated.py [hunk 20]
```diff
@@ -111,11 +117,11 @@ class RiRoIAlignRotated(nn.Module):
     """
 
     def __init__(self,
-                 out_size,
-                 spatial_scale,
-                 num_samples=0,
-                 num_orientations=8,
-                 clockwise=False):
+                 out_size: tuple,
+                 spatial_scale: int,
```
Comment (zhouzaida) [inline mmcv/ops/riroi_align_rotated.py:121]:
```suggestion
                 spatial_scale: float,
```

File: mmcv/ops/roi_align.py [hunk 21]
```diff
@@ -14,8 +16,9 @@
 class RoIAlignFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
-                 pool_mode, aligned):
+    def symbolic(g: torch._C.Graph, input: torch.Tensor, rois: torch.Tensor,
+                 output_size: tuple, spatial_scale: float, sampling_ratio: int,
+                 pool_mode: str, aligned: bool):
```
Comment (zhouzaida) [inline mmcv/ops/roi_align.py:21]:
```suggestion
    def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
                 pool_mode, aligned):
```

File: mmcv/ops/roi_align_rotated.py [hunk 22]
```diff
@@ -12,8 +15,9 @@
 class RoIAlignRotatedFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
-                 aligned, clockwise):
+    def symbolic(g: torch._C.Graph, input: torch.Tensor, rois: torch.Tensor,
+                 output_size: Union[int, tuple], spatial_scale: float,
+                 sampling_ratio: int, aligned: bool, clockwise: bool):
```
Comment (zhouzaida) [inline mmcv/ops/roi_align_rotated.py:20]:
```suggestion
    def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
                 aligned, clockwise):
```

File: mmcv/ops/roi_pool.py [hunk 23]
```diff
@@ -14,7 +16,8 @@
 class RoIPoolFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, rois, output_size, spatial_scale):
+    def symbolic(g: torch._C.Graph, input: torch.Tensor, rois: torch.Tensor,
+                 output_size: Union[int, tuple], spatial_scale: float):
```
Comment (zhouzaida) [inline mmcv/ops/roi_pool.py:20]:
```suggestion
    def symbolic(g, input, rois, output_size, spatial_scale):
```

File: mmcv/ops/rotated_feature_align.py [hunk 24]
```diff
@@ -21,7 +23,9 @@ class RotatedFeatureAlignFunction(Function):
     """
 
     @staticmethod
-    def symbolic(g, features, best_rbboxes, spatial_scale, points):
+    def symbolic(g: torch._C.Graph, features: torch.Tensor,
+                 best_rbboxes: torch.Tensor, spatial_scale: float,
+                 points: int):
```
Comment (zhouzaida) [inline mmcv/ops/rotated_feature_align.py:28]:
```suggestion
    def symbolic(g, features, best_rbboxes, spatial_scale, points):
```

File: mmcv/ops/rotated_feature_align.py [hunk 25]
```diff
@@ -84,9 +89,9 @@ def backward(ctx, grad_output):
...
 
 
-def rotated_feature_align(features,
-                          best_rbboxes,
-                          spatial_scale=1 / 8,
-                          points=1):
+def rotated_feature_align(features: torch.Tensor,
+                          best_rbboxes: torch.Tensor,
+                          spatial_scale: float = 1 / 8,
+                          points: int = 1):
```
Comment (zhouzaida) [inline mmcv/ops/rotated_feature_align.py:95]:
```suggestion
                          points: int = 1) -> torch.Tensor:
```

File: mmcv/ops/sync_bn.py [hunk 26]
```diff
@@ -19,8 +21,10 @@
 class SyncBatchNormFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, running_mean, running_var, weight, bias, momentum,
-                 eps, group, group_size, stats_mode):
+    def symbolic(g: torch._C.Graph, input: torch.Tensor,
+                 running_mean: torch.Tensor, running_var: torch.Tensor,
+                 weight: torch.Tensor, bias: torch.Tensor, momentum: float,
+                 eps: float, group: int, group_size: int, stats_mode: str):
```
Comment (zhouzaida) [inline mmcv/ops/sync_bn.py:27]:
```suggestion
    def symbolic(g, input, running_mean, running_var, weight, bias, momentum,
                 eps, group, group_size, stats_mode):
```

File: mmcv/ops/roi_align.py [hunk 27]
```diff
@@ -62,14 +64,14 @@ def symbolic(g, input, rois, output_size, spatial_scale, sampling_ratio,
...
-                spatial_scale=1.0,
-                sampling_ratio=0,
-                pool_mode='avg',
-                aligned=True):
+    def forward(ctx: Any,
+                input: torch.Tensor,
+                rois: torch.Tensor,
+                output_size: int,
+                spatial_scale: float = 1.0,
+                sampling_ratio: int = 0,
```
Comment (zhouzaida) [inline mmcv/ops/roi_align.py:72]:
```suggestion
                sampling_ratio: float = 0,
```

File: mmcv/ops/psa_mask.py [hunk 28]
```diff
@@ -12,15 +15,16 @@
 class PSAMaskFunction(Function):
 
     @staticmethod
-    def symbolic(g, input, psa_type, mask_size):
+    def symbolic(g, input: torch.Tensor, psa_type: str, mask_size: int):
```
Comment (zhouzaida) [inline mmcv/ops/psa_mask.py:18]:
```suggestion
    def symbolic(g, input, psa_type, mask_size):
```

File: mmcv/ops/roipoint_pool3d.py [hunk 29]
```diff
@@ -1,3 +1,6 @@
+from typing import Any
```
Comment (zhouzaida) [inline mmcv/ops/roipoint_pool3d.py:1]:
```suggestion
from typing import Any, Tuple
```

File: mmcv/ops/roipoint_pool3d.py [hunk 30]
```diff
@@ -17,11 +20,12 @@ class RoIPointPool3d(nn.Module):
...
     """
 
-    def __init__(self, num_sampled_points=512):
+    def __init__(self, num_sampled_points: int = 512):
         super().__init__()
         self.num_sampled_points = num_sampled_points
 
-    def forward(self, points, point_features, boxes3d):
+    def forward(self, points: torch.Tensor, point_features: torch.Tensor,
+                boxes3d: torch.Tensor) -> tuple:
```
Comment (zhouzaida) [inline mmcv/ops/roipoint_pool3d.py:28]:
```suggestion
                boxes3d: torch.Tensor) -> Tuple[torch.Tensor]:
```

File: mmcv/ops/roipoint_pool3d.py [hunk 31]
```diff
@@ -41,7 +45,11 @@ def forward(self, points, point_features, boxes3d):
 class RoIPointPool3dFunction(Function):
 
     @staticmethod
-    def forward(ctx, points, point_features, boxes3d, num_sampled_points=512):
+    def forward(ctx: Any,
+                points: torch.Tensor,
+                point_features: torch.Tensor,
+                boxes3d: torch.Tensor,
+                num_sampled_points: int = 512) -> tuple:
```
Comment (zhouzaida) [inline mmcv/ops/roipoint_pool3d.py:52]:
```suggestion
                num_sampled_points: int = 512) -> Tuple[torch.Tensor]:
```

File: mmcv/ops/scatter_points.py [hunk 32]
```diff
@@ -1,4 +1,6 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+from typing import Any, List, Optional
```
Comment (zhouzaida) [inline mmcv/ops/scatter_points.py:2]:
```suggestion
from typing import Any, List, Optional, Tuple
```

File: mmcv/ops/scatter_points.py [hunk 33]
```diff
@@ -14,7 +16,10 @@
 class _DynamicScatter(Function):
 
     @staticmethod
-    def forward(ctx, feats, coors, reduce_type='max'):
+    def forward(ctx: Any,
+                feats: torch.Tensor,
+                coors: torch.Tensor,
+                reduce_type: str = 'max') -> tuple:
```
Comment (zhouzaida) [inline mmcv/ops/scatter_points.py:22]:
```suggestion
                reduce_type: str = 'max') -> Tuple[torch.Tensor]:
```

File: mmcv/ops/scatter_points.py [hunk 34]
```diff
@@ -73,14 +80,16 @@ class DynamicScatter(nn.Module):
...
+                 average_points: bool):
         super().__init__()
 
         self.voxel_size = voxel_size
         self.point_cloud_range = point_cloud_range
         self.average_points = average_points
 
-    def forward_single(self, points, coors):
+    def forward_single(self, points: torch.Tensor,
+                       coors: torch.Tensor) -> tuple:
```
Comment (zhouzaida) [inline mmcv/ops/scatter_points.py:92]:
```suggestion
                       coors: torch.Tensor) -> Tuple[torch.Tensor]:
```

File: mmcv/ops/sparse_functional.py [hunk 35]
```diff
@@ -25,8 +27,10 @@ class SparseConvFunction(Function):
     """
 
     @staticmethod
-    def forward(ctx, features, filters, indice_pairs, indice_pair_num,
-                num_activate_out):
+    def forward(ctx: Any, features: torch.Tensor,
+                filters: torch.nn.parameter.Parameter,
```
Comment (zhouzaida) [inline mmcv/ops/sparse_functional.py:31]:
```suggestion
                filters: torch.nn.Parameter,
```

File: mmcv/ops/sparse_functional.py [hunk 36]
```diff
@@ -56,8 +60,10 @@ def backward(ctx, grad_output):
 class SparseInverseConvFunction(Function):
 
     @staticmethod
-    def forward(ctx, features, filters, indice_pairs, indice_pair_num,
-                num_activate_out):
+    def forward(ctx: Any, features: torch.Tensor,
+                filters: torch.nn.parameter.Parameter,
```
Comment (zhouzaida) [inline mmcv/ops/sparse_functional.py:64]:
```suggestion
                filters: torch.nn.Parameter,
```

File: mmcv/ops/sparse_functional.py [hunk 37]
```diff
@@ -87,8 +93,10 @@ def backward(ctx, grad_output):
 class SubMConvFunction(Function):
 
     @staticmethod
-    def forward(ctx, features, filters, indice_pairs, indice_pair_num,
-                num_activate_out):
+    def forward(ctx: Any, features: torch.Tensor,
+                filters: torch.nn.parameter.Parameter,
```
Comment (zhouzaida) [inline mmcv/ops/sparse_functional.py:97]:
```suggestion
                filters: torch.nn.Parameter,
```

File: mmcv/ops/sparse_modules.py [hunk 38]
```diff
@@ -118,14 +120,14 @@ def __len__(self):
     def sparity_dict(self):
         return self._sparity_dict
 
-    def add(self, module, name=None):
+    def add(self, module: Any, name: Optional[str] = None):
```
Comment (zhouzaida) [inline mmcv/ops/sparse_modules.py:123]:
```suggestion
    def add(self, module: Any, name: Optional[str] = None) -> None:
```

File: mmcv/ops/sparse_structure.py [hunk 39]
```diff
@@ -21,18 +24,18 @@ def scatter_nd(indices, updates, shape):
...
+                 batch_size: int,
+                 grid: Optional[torch.Tensor] = None):
         self.features = features
         self.indices = indices
         if self.indices.dtype != torch.int32:
             self.indices.int()
         self.spatial_shape = spatial_shape
         self.batch_size = batch_size
-        self.indice_dict = {}
+        self.indice_dict: Dict = {}
```
Comment (zhouzaida) [inline mmcv/ops/sparse_structure.py:38]:
```suggestion
        self.indice_dict: dict = {}
```

File: mmcv/ops/upfirdn2d.py [hunk 40]
```diff
@@ -264,14 +274,14 @@ def upfirdn2d(input, kernel, up=1, down=1, pad=(0, 0)):
...
-        out = upfirdn2d_native(input, kernel, up[0], up[1], down[0], down[1],
-                               pad[0], pad[1], pad[2], pad[3])
+            pad = (pad[0], pad[1], pad[0], pad[1])  # type: ignore
+
+        out = upfirdn2d_native(input, kernel,
+                               to_2tuple(up)[0],
+                               to_2tuple(up)[1],
+                               to_2tuple(down)[0],
+                               to_2tuple(down)[1], pad[0], pad[1], pad[2],
+                               pad[3])
```
Comment (zhouzaida) [inline mmcv/ops/upfirdn2d.py:284]:
```suggestion
        _up = to_2tuple(up)

        _down = to_2tuple(down)

        out = upfirdn2d_native(input, kernel, _up[0], _up[1], _down[0],
                               _down[1], pad[0], pad[1], pad[2], pad[3])
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**